flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/SourceStreamTaskTest.java

```diff
@@ -856,4 +887,41 @@ private void output(String record) {
             output.collect(new StreamRecord<>(record));
         }
     }
+
+    /**
+     * This source sleeps a little bit before processing cancellation and records whether it was
+     * interrupted by the {@link SourceStreamTask} or not.
+     */
+    private static class WasInterruptedTestingSource implements SourceFunction<String> {
+        private static final int MAX_POST_CANCEL_ITERATIONS = 100;
+        private static final long serialVersionUID = 1L;
+        private volatile boolean running;
+
+        public static boolean wasInterrupted;
+
+        @Override
+        public void run(SourceContext<String> ctx) throws Exception {
+            wasInterrupted = false;
+            try {
+                int postCancelIterations = 0;
+                while (running || postCancelIterations < MAX_POST_CANCEL_ITERATIONS) {
```

Comment:
I think it's better to signal explicitly that this thread can exit. Otherwise, if there is a pause between `cancel` call and (errouneous) `interrupt` that interrupt can go unnoticed.

I reverted the fix and added 1s pause in `SourceStreamTask` - and the test passed (with running=true)

Reply:
Yes, even without the sleep test can have false negative from time to time. Since this require unexpected sleep/hiccup of 100ms (or more), it looks like chances of this happening are pretty low (`<10%` ?), so the test would catch a regression quite quickly. This test checks:
1. we have a source that ignores for `x` milliseconds cancellation
2. checks that within those `x` ms source thread was not interrupted. 

`notifyCheckpointCompleteAsync`? Did you mean `StreamTask#notifyCheckpointComplete` -> `CheckpointListener#notifyCheckpointComplete`? 

Exiting on `CheckpointListener#notifyCheckpointComplete` would still require a sleep to check if the interruption was not called after it. Also in the stop with savepoint case, both `CheckpointListener#notifyCheckpointComplete` and `SourceFunction#cancel` are both being called from `StreamTask#notifyCheckpointComplete` one after another. As a matter of fact, `notifyCheckpointComplete()` is called before `cancel()`, so it would require even so slightly longer/more `MAX_POST_CANCEL_ITERATIONS`.

I don't see how can we reliably test this without some sleep based probability window, because for that to be possible we would need to have a notification hook that is always triggered **after** interruption, so that the source would know if the interruption was supposed to happen it would have already happened. But we don't have anything like that.

Reply:
I mean't to add in `testStopWithSavepointShouldNotInterruptTheSource` after
```
Future<Void> notifyFuture =
    harness.streamTask.notifyCheckpointCompleteAsync(checkpointId);```
```
something like this:
```
while (harness.streamTask.isMailboxLoopRunning() && !notifyFuture.isDone()) {
    harness.streamTask.runMailboxStep();
}
notifyFuture.get();

// interruption, if any, must have happened before
WasInterruptedTestingSource.allowExit();
```
This allows to detect otherwise undetected interrupt (ignore `interrupt` flag in `cancelTask()` and sleep 1s).
As it allows to drop the loop in the source, complexity seems the same or a bit lower.

Reply:
Good idea, thanks for the suggestion :)